applications/tari_base_node/src/builder.rs

```diff
@@ -495,43 +493,155 @@ where
+                                    OutputManagerEvent::UtxoValidationFailure(s) => {
+                                        trace!(
+                                            target: LOG_TARGET,
+                                            "UTXO validation event 'UtxoValidationFailure' ({}), restarting.", s,
+                                        );
+                                        break;
+                                    },
+                                    // Exit upon success
+                                    OutputManagerEvent::UtxoValidationSuccess(s) => {
+                                        trace!(
+                                            target: LOG_TARGET,
+                                            "UTXO validation event 'UtxoValidationSuccess' ({}).", s,
+                                        );
+                                        if validation == ValidationTask::InvalidOutputs {
+                                            validation = ValidationTask::UnspentOutputs;
+                                        } else {
+                                            end_protocol = true;
+                                        }
+                                        break;
+                                    },
+                                    // Wait for the next event if timed out (several can be attempted)
+                                    OutputManagerEvent::UtxoValidationTimedOut(s) => {
+                                        trace!(
+                                            target: LOG_TARGET,
+                                            "UTXO validation event 'UtxoValidationTimedOut' ({}), waiting.", s,
+                                        );
+                                        continue;
+                                    }
+                                    // Wait for the next event upon an error
+                                    OutputManagerEvent::Error(s) => {
+                                        trace!(
+                                            target: LOG_TARGET,
+                                            "UTXO validation event 'Error({})', waiting.", s,
+                                        );
+                                        continue;
+                                    },
+                                    // Wait for the next event upon anything else
+                                    _ => {
+                                        trace!(
+                                            target: LOG_TARGET,
+                                            "UTXO validation unknown event, waiting.",
+                                        );
+                                        continue;
+                                    },
+                                }
+                            },
+                            // Restart the protocol if it timed out
+                            () = delay => {
```

Comment:
This is actually unnecessary, the UTXO validation protocols will timeout on their own and you can do what you do with this branch above in the `OutputManagerEvent::UtxoValidationTimedOut =>` branch

Reply:
This is a safety net. In one of my system-level tests, the protocol never returned with another event and the task got stuck in `futures::select! {`

Reply:
It should not. The validation protocols have a built in timeout and the issue is you are not using the instance of the validations protocol timeout.

Reply:
I would feel uneasy about removing the safety net. The `OutputManagerEvent::UtxoValidationTimedOut` event is used in the logic, see below.